plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/collector/StorPoolAbandonObjectsCollector.java

```diff
@@ -0,0 +1,320 @@
+import com.cloud.utils.concurrency.NamedThreadFactory;
+import com.cloud.utils.db.DB;
+import com.cloud.utils.db.Transaction;
+import com.cloud.utils.db.TransactionCallbackNoReturn;
+import com.cloud.utils.db.TransactionLegacy;
+import com.cloud.utils.db.TransactionStatus;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+
+public class StorPoolAbandonObjectsCollector extends ManagerBase implements Configurable {
+    private static Logger log = Logger.getLogger(StorPoolAbandonObjectsCollector.class);
+    @Inject
+    private PrimaryDataStoreDao storagePoolDao;
+    @Inject
+    private StoragePoolDetailsDao storagePoolDetailsDao;
+
+    private ScheduledExecutorService _volumeTagsUpdateExecutor;
+    private static final String ABANDON_LOG = "/var/log/cloudstack/management/storpool-abandoned-objects";
+
+
+    static final ConfigKey<Integer> volumeCheckupTagsInterval = new ConfigKey<Integer>("Advanced", Integer.class,
+            "storpool.volume.tags.checkup", "86400",
+            "Minimal interval (in seconds) to check and report if StorPool volume exists in CloudStack volumes database",
+            false);
+    static final ConfigKey<Integer> snapshotCheckupTagsInterval = new ConfigKey<Integer>("Advanced", Integer.class,
+            "storpool.snapshot.tags.checkup", "86400",
+            "Minimal interval (in seconds) to check and report if StorPool snapshot exists in CloudStack snapshots database",
+            false);
+
+    @Override
+    public String getConfigComponentName() {
+        return StorPoolAbandonObjectsCollector.class.getSimpleName();
+    }
+
+    @Override
+    public ConfigKey<?>[] getConfigKeys() {
+        return new ConfigKey<?>[] { volumeCheckupTagsInterval, snapshotCheckupTagsInterval };
+    }
+
+    @Override
+    public boolean start() {
+        init();
+        return true;
+    }
+
+    private void init() {
+        _volumeTagsUpdateExecutor = Executors.newScheduledThreadPool(2,
+                new NamedThreadFactory("StorPoolAbandonObjectsCollector"));
+        StorPoolHelper.appendLogger(log, ABANDON_LOG, "abandon");
+        if (volumeCheckupTagsInterval.value() > 0) {
+            _volumeTagsUpdateExecutor.scheduleAtFixedRate(new StorPoolVolumesTagsUpdate(),
+                    volumeCheckupTagsInterval.value(), volumeCheckupTagsInterval.value(), TimeUnit.SECONDS);
+        }
+        if (snapshotCheckupTagsInterval.value() > 0) {
+            _volumeTagsUpdateExecutor.scheduleAtFixedRate(new StorPoolSnapshotsTagsUpdate(),
+                    snapshotCheckupTagsInterval.value(), snapshotCheckupTagsInterval.value(), TimeUnit.SECONDS);
+        }
+    }
+
+    class StorPoolVolumesTagsUpdate extends ManagedContextRunnable {
+
+        @Override
+        @DB
+        protected void runInContext() {
+            List<StoragePoolVO> spPools = storagePoolDao.findPoolsByProvider(StorpoolUtil.SP_PROVIDER_NAME);
+            if (spPools != null && spPools.size() > 0) {
```

Comment:
We could invert this condition (using `CollectionUtils#isEmpty`) and add a return to reduce the indentation of the block.